active/0014-rolling-cluster-upgrade.md

```diff
@@ -0,0 +1,122 @@
+
+## Abstract
+
+Currently EMQ X upgrade procedure has two modes of upgrade:
+
+1. Patch releases.
+   Upgrading between the patch versions is done by patching the live nodes using Erlang hot code patching feature.
+   The patch is loaded to the existing running nodes.
+   This is a low-maintenance upgrade path, however severely limited in the scope of the changes.
+1. Minor and major releases requires taking down the entire cluster and redirecting the MQTT traffic to a different cluster running the new version of the software.
+   This approach introduces a lot of operational overhead.
+
+EMQ X should support rolling upgrade of the cluster when nodes (or pods) are taken down and replaced one by one for minor version upgrades.
+Patch version upgrades will remain the same.
+
+This EIP introduces the guidelines for writing the backward- and forward-compatible code.
+Also it documents the necessary changes to the existing code base.
+
+## Motivation
+
+In order to make upgrading the cluster smoother, EMQ X should support rolling cluster upgrades and green-blue deployments.
+It should be possible to upgrade the cluster without taking it down entirely.
+
+## Design
+
+Live upgrade paths will be limited to `major.minor.* -> major.minor+1.*` formula.
+
+A new concept of a "backplane API" is introduced.
+
+Cluster upgrade should be split in roughly three stages:
+
+1. Optional step: inject the forward compatibility support into the old upgrade.
+   This may be needed because we can't predict in advance what will be changed in the next release.
+   The upgrade code should be able to inject pre-upgrade beams to the old release.
+   This part of the upgrade procedure should be idempotent and reversible.
+   It will work much like a patch version upgrade.
+1. Rolling upgrade of the cluster that involves taking nodes down and replacing them with the newer version.
+   Both versions of the code work together in the same cluster.
+   This part of the upgrade can be revesible.
+1. Once all the nodes are upgraded, a data migration can start.
+   This part of the upgrade is not reversible.
+
+Deprecated APIs can be removed in the next release.
+
+There are two major areas that need to be considered to support this kind of upgrade:
+
+1. RPC compatibility
+1. Mnesia schema backward-compatibility
+
+A helper mnesia table tracking the upgrade state and release version for each node can be used to perform the checks between proceeding to the next stage of the upgrade.
+
+### RPC compatibility
+
+In order to simplify the reasoning about the backplane API backward-compatibility, all the functions that may be called remotely should be identified and gathered in specialized modules.
+Static checks should be introduced to make sure these functions are not removed without being deprecated first.
+(Perhaps dump the list of these functions to a file that is held in the repo and used in the CI?)
+Direct sending messages to the remote processes should be prohibited.
```

Comment:
hopefully, there is no 3rd party app doing this.

Reply:
There could be things like that... It means we should prohibit (or limit) updating of the dependencies that don't comply with the BCBPAPI (backward-compatibly backplane API) standards.